ocis-pkg/runner/grouprunner.go

```diff
@@ -0,0 +1,130 @@
+package runner
+
+import (
+	"context"
+)
+
+// GroupRunner represent a group of tasks that need to run together.
+// The expectation is that all the tasks will run at the same time, and when
+// one of them stops, the rest will also stop.
+//
+// The GroupRunner is intended to be used to run multiple services, which are
+// more or less independent from eachother, but at the same time it doesn't
+// make sense to have any of them stopped while the rest are running.
+// Basically, either all of them run, or none of them.
+// For example, you can have a GRPC and HTTP servers running, each of them
+// providing a piece of functionality, however, if any of them fails, the
+// feature provided by them would be incomplete or broken.
+//
+// Note that, as services, the task aren't expected to stop by default.
+// This means that, if a task finishes naturally, the rest of the task will
+// asked to stop as well.
+type GroupRunner struct {
+	runners []*Runner
+}
+
+// NewGroupRunner will create a GroupRunner
+func NewGroupRunner() *GroupRunner {
+	return &GroupRunner{
+		runners: []*Runner{},
+	}
+}
+
+// Add will add a runner to the group.
+//
+// It's mandatory that each runner in the group has an unique id, otherwise
+// there will be issues
+func (gr *GroupRunner) Add(r *Runner) {
+	gr.runners = append(gr.runners, r)
+}
+
+// Run will execute all the tasks in the group at the same time.
+//
+// Similarly to the "regular" runner's `Run` method, the execution thread
+// will be blocked here until all tasks are completed, and their results
+// will be available (each result will have the runner's id so it's easy to
+// find which one failed). Note that there is no guarantee about the result's
+// order, so the first result in the slice might or might not be the first
+// result to be obtained.
+//
+// When the context is marked as done, the groupRunner will call all the
+// stoppers for each runner to notify each task to stop. Note that the tasks
+// might still take a while to complete.
+//
+// If a task finishes naturally (with the context still "alive"), it will also
+// cause the groupRunner to call the stoppers of the rest of the tasks. So if
+// a task finishes, the rest will also finish.
+// Note that it is NOT expected for the finished task's stopper to be called
+// in this case.
+func (gr *GroupRunner) Run(ctx context.Context) []*Result {
+	results := make(map[string]*Result)
+
+	ch := make(chan *Result, len(gr.runners)) // no need to block writing results
+	for _, runner := range gr.runners {
+		runner.RunNoContext(ch)
```

Comment:
The `Run` method will block the goroutine, so we'd need to spawn new goroutines in order not to block the current one. I think this would lead to a 2x+1 goroutines running instead of x+1 (being x the number of runners in the group)
In addition, the `Run` method (for both the `Runner` and `GroupRunner`) is expected to run synchronously, so we'll need a way to wait for the spawned goroutines to finish.

With the current code, the `RunNoContext` will spawn the goroutines (one per runner) and they'll leave their results in the channel, so the current goroutine can wait in the channel.

Maybe changing method name to `RunAsync` makes more sense.